google/resource_bigtable_gc_policy.go

```diff
@@ -0,0 +1,227 @@
+package google
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+	"time"
+
+	"cloud.google.com/go/bigtable"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+const (
+	GCPolicyModeIntersection = "INTERSECTION"
+	GCPolicyModeUnion        = "UNION"
+)
+
+func resourceBigtableGCPolicy() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceBigtableGCPolicyCreate,
+		Read:   resourceBigtableGCPolicyRead,
+		Delete: resourceBigtableGCPolicyDestroy,
+
+		Schema: map[string]*schema.Schema{
+			"instance_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"table": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"family": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"mode": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{GCPolicyModeIntersection, GCPolicyModeUnion}, false),
+			},
+
+			"max_age": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"days": {
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+					},
+				},
+			},
+
+			"max_version": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"number": {
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+					},
+				},
+			},
+
+			"project": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceBigtableGCPolicyCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	ctx := context.Background()
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	instanceName := d.Get("instance_name").(string)
+	c, err := config.bigtableClientFactory.NewAdminClient(project, instanceName)
+	if err != nil {
+		return fmt.Errorf("Error starting admin client. %s", err)
+	}
+
+	defer c.Close()
+
+	gcPolicy, err := generateBigtableGCPolicy(d)
+	if err != nil {
+		return err
+	}
+
+	tableName := d.Get("table").(string)
+	family := d.Get("family").(string)
+
+	if err := c.SetGCPolicy(ctx, tableName, family, gcPolicy); err != nil {
+		return err
+	}
+
+	table, err := c.TableInfo(ctx, tableName)
+	if err != nil {
+		return fmt.Errorf("Error retrieving table. Could not find %s in %s. %s", tableName, instanceName, err)
+	}
+
+	for _, i := range table.FamilyInfos {
+		if i.Name == family {
+			d.SetId(i.GCPolicy)
+		}
+	}
+
+	return resourceBigtableGCPolicyRead(d, meta)
+}
+
+func resourceBigtableGCPolicyRead(d *schema.ResourceData, meta interface{}) error {
```

Comment:
Is it possible to recreate a GCPolicy from the string in FamilyInfo? https://github.com/terraform-providers/terraform-provider-google/blob/ab7c29f09dafc6d836888b3db57b53f148e37727///vendor/cloud.google.com/go/bigtable/admin.go#L231

Drift detection would be nice if it can, but it's not a must-have.

Reply:
It's weird I forget about this, you mean about the SetId("...") to ensure nothing change between two plan/apply? If yes, should be good now

Reply:
The `id` value won't actually change anything, so not setting it is fine. I'm curious whether we can extract values like `max_age` from the GCPolicy string, so we can set those in state.

Reply:
I see, it's probably possible, but I thinking about all cases, what if Google change something? Or add another policy? I am not even sure order of the policy is guaranteed (can be max_age && max_version / max_version && max_age)

I wonder, what's happen in terraform if you have 2 resources, with the same ID (actually, that can happen if you set same gc policy on two table), should I prefix ID with the column name?

Reply:
2 resources can share an id, Terraform only distinguishes between the `id` value being unset and set.